File: packages/totem-sdk/packages/manifest/src/id.ts
/**
 * computeManifestId — deterministic stable ID for any manifest.
 *
 * The ID is SHA3-256 over a type-specific stable key string.
 * It must be stable across version bumps (version is NOT part of the input).
 *
 * Stable key rules:
 *   AppManifest        → "app"        + authorAddress + pearTopicKey
 *   CapabilityManifest → "capability" + agentAddress  + capabilityName
 *   DAppManifest       → "dapp"       + authorAddress + contractHash
 *   EdgeServiceManifest → "edge-service" + operatorAddress + serviceType + name
 */

import { sha3_256 } from '@noble/hashes/sha3.js';
import type { Manifest } from './types.js';

function bytesToHex(b: Uint8Array): string {
  return Array.from(b)
    .map((x) => x.toString(16).padStart(2, '0'))
    .join('');
}

export function computeManifestId(manifest: Manifest): string {
  let stableKey: string;

  switch (manifest.type) {
    case 'app':
      stableKey = `app\0${manifest.authorAddress}\0${manifest.pearTopicKey}`;
      break;
    case 'capability':
      stableKey = `capability\0${manifest.agentAddress}\0${manifest.capabilityName}`;
      break;
    case 'dapp':
      stableKey = `dapp\0${manifest.authorAddress}\0${manifest.contractHash}`;
      break;
    case 'edge-service':
      stableKey = `edge-service\0${manifest.operatorAddress}\0${manifest.serviceType}\0${manifest.name}`;
      break;
    default: {
      const _exhaustive: never = manifest;
      throw new Error(`computeManifestId: unknown manifest type: ${JSON.stringify(_exhaustive)}`);
    }
  }

  return bytesToHex(sha3_256(new TextEncoder().encode(stableKey)));
}

File: packages/totem-sdk/packages/manifest/rust/src/id.rs
use sha3::{Digest, Sha3_256};

use crate::canonical::to_hex;

pub fn compute_manifest_id(manifest: &serde_json::Value) -> Result<String, String> {
    let manifest_type = manifest["type"].as_str().unwrap_or("");
    let stable_key = match manifest_type {
        "app" => {
            let author = manifest["authorAddress"].as_str().unwrap_or("");
            let pear = manifest["pearTopicKey"].as_str().unwrap_or("");
            format!("app\0{}\0{}", author, pear)
        }
        "capability" => {
            let agent = manifest["agentAddress"].as_str().unwrap_or("");
            let name = manifest["capabilityName"].as_str().unwrap_or("");
            format!("capability\0{}\0{}", agent, name)
        }
        "dapp" => {
            let author = manifest["authorAddress"].as_str().unwrap_or("");
            let hash = manifest["contractHash"].as_str().unwrap_or("");
            format!("dapp\0{}\0{}", author, hash)
        }
        "edge-service" => {
            let operator = manifest["operatorAddress"].as_str().unwrap_or("");
            let svc_type = manifest["serviceType"].as_str().unwrap_or("");
            let name = manifest["name"].as_str().unwrap_or("");
            format!("edge-service\0{}\0{}\0{}", operator, svc_type, name)
        }
        other => return Err(format!("unknown manifest type: {}", other)),
    };

    let mut hasher = Sha3_256::new();
    hasher.update(stable_key.as_bytes());
    Ok(to_hex(&hasher.finalize()))
}
